src/OnlineSales/Tasks/SyncEmailLogTask.cs

```diff
@@ -44,15 +48,44 @@ public async override Task<bool> Execute(TaskExecutionLog currentJob)
         {
             try
             {
-                var maxId = await logService.GetMaxId(SourceName);
-                var batch = await dbContext.EmailLogs!.Where(e => e.Id > maxId).OrderBy(e => e.Id).Take(batchSize).ToArrayAsync();
-                var tasks = batch.Select(Convert).ToArray();
-                Task.WaitAll(tasks);
-                await dbContext.SaveChangesAsync();
-                var res = await logService.AddActivityRecords(tasks.Select(t => t.Result).ToList());
-                if (!res)
+                using (var scope = provider.CreateScope())
                 {
-                    throw new SyncEmailLogTaskException("Unable to log email events");
+                    var contactService = scope.ServiceProvider.GetRequiredService<ContactService>();
+
+                    var maxId = await logService.GetMaxId(SourceName);
+                    var batch = await dbContext.EmailLogs!.Where(e => e.Id > maxId).OrderBy(e => e.Id).Take(batchSize).ToListAsync();
+                    var batchContacts = batch.Select(b => b.Recipient).ToArray();
+                    var existingContacts = await dbContext.Contacts!.Where(contact => batchContacts.Contains(contact.Email)).ToDictionaryAsync(k => k.Email);
+
+                    var converted = batch.Select(log =>
+                    {
+                        Contact? contact;
+
+                        if (!existingContacts.TryGetValue(log.Recipient, out contact))
+                        {
+                            contact = contactService.SaveAsync(new Contact
```

Comment:
@Exmirai .Result.Entity will be the same instance of the object which you constructed yourselves and passed to SaveAsync. What is the reason for the complexity you introduced with the returning of EntityEntry from the SaveAsync method? Please create and object first and save it into a local variable and than use that variable where needed. There is no need to return EntityEntry.